Use Python types to declare document fields (#1845)
elastic/elasticsearch-dsl-py: docs/persistence.rst

    # empty range is unbounded
    Range().lower # None, False

Python Type Hints
~~~~~~~~~~~~~~~~~

Document fields can be defined using standard Python type hints if desired.
Here are some simple examples:

.. code:: python

    from typing import Optional

    class Post(Document):
        title: str                      # same as Text(required=True)
        created_at: Optional[datetime]  # same as Date(required=False)

>         published: bool                 # same as Boolean(required=True)    

Reviewer: Would it be helpful to call out that using type hints has different syntax than assignments somewhere?

Author: There is an error at runtime if you mistakenly use a DSL type as a type hint. If you do this:

```python
class MyDoc(Document):
    title: Text
```

You get this error:

```
TypeError: Cannot map type <class 'elasticsearch_dsl.field.Text'>
```

And if you do this:

```python
class MyDoc(Document):
    title: Text(required=True)
```

The error is:

```
TypeError: Cannot map type Text()
```

Python types are mapped to their corresponding field type according to the
following table:

.. list-table:: Python type to DSL field mappings
   :header-rows: 1

   * - Python type
     - DSL field
   * - ``str``
     - ``Text(required=True)``
   * - ``bool``
     - ``Boolean(required=True)``
   * - ``int``
     - ``Integer(required=True)``
   * - ``float``
     - ``Float(required=True)``
   * - ``bytes``
     - ``Binary(required=True)``
   * - ``datetime``
     - ``Date(required=True)``
   * - ``date``

>      - ``Date(format="yyyy-MM-dd", required=True)``

Reviewer: This is very helpful, should we also add List and Optional?

Author: Good call, I'll add them to the table.

Reviewer: Did you decide against adding `Optional`?

Author: I didn't. Sorry, I added List, but left out `Optional`. I'm adding it.

Author: Okay, so I tried a couple of different formats and found that growing the table makes it too long and more difficult to visually parse. So in the end I kept the table small with just the required variants of the Python native types, and right after the table I show examples of `Optional`, `List`, `Object` and `Nested`. Let me know what you think.

        address: Address         # same as Object(Address)
        comments: List[Comment]  # same as Nested(Comment)

Unfortunately it is impossible to have Python type hints that uniquely
identify every possible Elasticsearch field type. To choose a field type that
is different thant the ones in the table above, the field instance can be added
explicitly as a right-side assignment in the field declaration. The next
example creates a field that is typed as ``str``, but is mapped to ``Keyword``
instead of ``Text``:

.. code:: python

    class MyDocument(Document):
        category: str = Keyword()

This form can also be used when additional options need to be given to
initialize the field, such as when using custom analyzer settings:

.. code:: python

    class Comment(InnerDoc):
        content: str = Text(analyzer='snowball')

The standard ``Optional`` modifier from the Python ``typing`` package can be
used to change a typed field from required to optional. The ``List`` modifier
can be added to a field to convert it to an array, similar to using the
``multi=True`` argument on the field object.

When using type hints as above, subclasses of ``Document`` and ``InnerDoc``

> inherit some of the behaviors associated with Python dataclasses. If

Reviewer: I was nodding along until here, and was surprised to see that subclasses using type hints "inherited some of the behaviors associated with Python dataclasses". We don't know which ones. If it's not important which behaviors, maybe we should not mention them at all? Or maybe link to `dataclass_transform` docs?

Author: I'll clarify this. For some reason I tried to avoid mentioning `dataclass_transform` here, but now that you make me think about this, it is best to be clear about what this means exactly.

> inherit some of the behaviors associated with Python dataclasses. If
> necessary, the ``mapped_field()`` wrapper can be used on the right side of a
> typed field declaration, enabling dataclass options such as ``default`` or
> ``default_factory`` to be included:

Reviewer: Starting with "If necessary" feels weird because I have no idea what's necessary or not. What do you think of this wording?

```suggestion
inherit some of the behaviors associated with Python dataclasses. To add
per-field dataclass options such as ``default`` or ``default_factory`` , the
``mapped_field()`` wrapper can be used on the right side of a typed field
declaration:
```

Also, maybe we should mention that this is similar to `dataclass.field()` (with an Intersphinx link) and maybe explain why `dataclass.field()` does not work here?

`pyright <https://github.com/microsoft/pyright>`_ can use the type hints and
the dataclass-specific options added to the ``mapped_field()`` function to
improve type inference and provide better real-time suggestions in IDEs.

One situation in which type checkers can't infer the correct type is when
using fields as class attributes. Consider the following example:

.. code:: python

    class MyDocument(Document):
        title: str = mapped_field(default="no title")

    doc = MyDocument()
    # doc.title is typed as "str" (correct)
    # MyDocument.title is also typed as "str" (incorrect)

To help type checkers correctly identify class attributes as such, the ``M``
generic must be used as a wrapper to the type hint, as shown in the next
example:

.. code:: python

    from elasticsearch_dsl import M

    class MyDocument(Document):
        title: M[str]
        created_at: M[datetime] = mapped_field(default_factory=datetime.now)

    doc = MyDocument()
    # doc.title is typed as "str"
    # MyDocument.title is typed as "InstrumentedField"

Note that the ``M`` type hint does not provide any runtime behavior, it just

> provides additional typing declarations for type checkers.

Reviewer: Should we encourage using it everywhere type hints are used? If not, then should it only be added to specific fields to fix typing issues? It feels some sort of short guideline would help here.

Author: I will clarify this. I think the guideline should be that `M` helps achieve zero errors/warnings for people who type check their code, especially if they use strict mode. And this is only for one use case, which is when you use class attributes to reference a field name. For any other usages, or if you don't care about type checkers, there is no need to complicate your code with it.

The ``InstrumentedField`` objects returned when fields are accessed as class
attributes are proxies for the field instances that can be used anywhere a
field needs to be referenced, such as when specifying sort options in a
``Search`` object:

.. code:: python

    # sort by creation date descending, and title ascending

>     s = MyDocument.search().sort(-MyDocument.created_at, MyDocument.title)

Reviewer: Are you saying that this sort() call would not be accepted by mypy without `M`?

Author: Yes, this is the only reason we need `M`, unfortunately. Without it, the class attributes are typed the same as the instance attributes.

        ...
    
    class Post(Document):
        address: Address         # same as address = Object(Address)
        comments: List[Comment]  # same as comments = Nested(Comment)

Unfortunately it is impossible to have Python type hints that uniquely
identify every possible Elasticsearch field type. To choose a field type that
is different than the ones in the table above, the field instance can be added
explicitly as a right-side assignment in the field declaration. The next
example creates a field that is typed as ``str``, but is mapped to ``Keyword``
instead of ``Text``:

.. code:: python

    class MyDocument(Document):
        category: str = Keyword(required=True)

This form can also be used when additional options need to be given to
initialize the field, such as when using custom analyzer settings:

.. code:: python

    class Comment(InnerDoc):

>         content: str = Text(analyzer='snowball', required=True)

Reviewer: Do we still need this part given the first example already has an option (`required=True`)?

Author: Okay, yes, I see that this makes less sense now. I'll convert the previous example to optional, so that this is the first time additional options are added.
